### ISSUE TITLE
Delta for adding TCP support?

### ISSUE THREAD

**Issue by samlevin** [deislabs/osiris#26]
**What would you like to be added?**
Support for TCP connections (instead of HTTP-only)

**Why is this needed?**
Interested in downscaling pods that accept TCP connection and don't make use of HTTP

I saw in another ticket (about gRCP and HTTP2) that Osiris relies on L7 proxies and thus only supports HTTP currently. What work would be needed in order to support TCP? Are there any known blockers or limitations that make it impossible altogether? If so, why/why not?

Assuming that it's possible, I'd be interested in helping out

**Comment by krancour** [comment]
@samlevin, I'm actively working on HTTP/2 support-- including both h2 (HTTP/2 over TLS) and h2c (HTTP/2 without TLS). The approach will amend the activator component to sniff connections to recognize them as either being used for plain text HTTP connections _or_ TLS-encrypted traffic, then dynamically choose to use an L7 or L4 proxy.

In the case of the L7 proxy, we can figure out which deployment to revive by looking at HTTP `Host` header. In the case of an L4 proxy, SNI headers give us (roughly) the same information during the TLS handshake.

For non-TLS-encrypted TCP traffic using unknown L7 protocols, the activator will not have the means to discern which deployment needs reviving. From my perspective, that would make generic TCP support all but impossible. The only way I could imagine that working would require an activator component per deployment, and at that point, you'll likely negate the cost savings of scaling those deployments to zero.

**Comment by samlevin** [comment]
That makes sense. Thanks for the detailed reply. On a related note - would it be possible to use Osiris' zeroscaler and metrics server solely for the purpose of downscaling TCP-based pods? Is it possible to increase the request count on any given request (via any protocol), or is that also specific to the L7 proxy/proxies used internally?

**Comment by krancour** [comment]
> would it be possible to use Osiris' zeroscaler and metrics server solely for the purpose of downscaling TCP-based pods?

If you're asking whether it would be possible to scale such deployments to zero without any concern for scaling them back up _from_ zero, it might theoretically be possible once HTTP/2 support is implemented.

Osiris' internal components each have a very narrow and discrete set of responsibilities. The zeroscaler watches deployments that are already scaled down to a configurable minimum (usually one) and tries to detect when all those pods are idling. When they are, it initiates the scale-to-zero process by changing the deployment's number of desired replicas. _That's it._

Reactivation comes courtesy of he activator component, but traffic for applications that are scaled to zero only ever _reaches_ the activator of the service that ordinarily selects those pods as endpoints is annotated as Osiris-enabled. Without that annotation, Osiris' endpoints controller will never divert such traffic to the activator.

So. It could theoretically be possible to let the zeroscaler do its thing and _not_ Osiris-enable your service. Frankly, this is a case I hadn't considered before now, so there might be some gotchas lurking in here. At any rate, it will be contingent on the HTTP/2 support being implemented, because until then, we're not doing _anything_ at L4.

> Is it possible to increase the request count on any given request (via any protocol), or is that also specific to the L7 proxy/proxies used internally?

An L4 proxy operates purely in terms of connections and bytes rather than in terms of requests-- which are a feature of certain L7 (application layer) protocols-- notably, HTTP/1.x. So the metrics actually need to be retooled to support anything at L4, including my work to support HTTP/2.

If you substitute the word "connection" for "request" in your question, the answer is yes, with a few caveats. That scheme wouldn't account for connections that stay open for a long time, which is possible with many protocols, including HTTP/2. Additional metrics and logic need to be introduced to detect those and count them as "activity" so that pods aren't misjudged to be idle just because one metric like "connection count" didn't increase over a given period. Again-- I expect that all to be accounted for when adding HTTP/2 support.

Bottom line is HTTP/2 support might, incidentally, get you pretty far toward what you're asking about. I'll try to remain mindful of this as I'm working on it, but I cannot make any guarantees as HTTP/2 is the real priority right now.

**Comment by samlevin** [comment]
Sounds great. I'll be watching from the sidelines. Thanks again for the info and best of luck in your efforts

**Comment by krancour** [comment]
I'm going to close this because plain TCP, at least w/out TLS+SNI, isn't something we can accomodate.

**Comment by ajbouh** [comment]
I'm looking to learn a bit more about what makes this unachievable without TLS+SNI. Is there a lurking assumption or implementation detail that's generating this constraint?

**Comment by krancour** [comment]
@ajbouh a TCP connection, by itself, doesn't include sufficient metadata for the activator to know which deployment needs to be scaled up from zero.

With HTTP, we get this information at L7 instead-- i.e. from the HTTP `Host` header. With HTTPS (HTTP over TLS), we can get this information from headers in the SNI extension to the TLS protocol. 

For all _other_ TCP connections, where the L7 protocol could be anything, we cannot know if sufficient information even exists to activate (and route traffic to) the correct deployment.

**Comment by ajbouh** [comment]
@krancour how does Osiris currently register itself to handle incoming HTTP traffic?

Perhaps we can listen on many host ports, one per TCP service? If we route each scaled-down service to a unique port, then we can use a simple lookup table to determine the correct deployment. Obviously there are some limits to this approach, but perhaps it would work well enough?

**Comment by krancour** [comment]
> how does Osiris currently register itself to handle incoming HTTP traffic?

Osiris is made up of many components. I suspect this question is really about how traffic is redirected to the activator?

There is an endpoints controller that, for each Osiris-enabled service, watches the pods for that service. When none are in a ready state, it substitutes activator endpoints for application endpoints.

It's probably relevant to discuss how the activator works today:

The activator maintains map of all the different ways an Osiris-enabled service can be addressed (e.g. private and public IPs, DNS names, etc.) to deployments that should be activated when a request it serves is addressed accordingly. The activator sniffs all connections to figure out whether they are HTTP or HTTPS (TCP+SNI, really) and based on which it is, it knows how to figure out what the real destination of the request is and that becomes the input to that deployment lookup.

> Perhaps we can listen on many host ports...

There's no reason to involve host ports in this. That one detail put aside, your idea isn't terrible.

> Perhaps we can listen on many ~~host~~ ports, one per TCP service? If we route each scaled-down service to a unique port, then we can use a simple lookup table to determine the correct deployment. 

If I may echo your proposal in more detail: The activator would listen on many ports. (Not one per Osiris-enabled service as you suggest, but actually one per _port_ of each Osiris-enabled service. So, for the sake of illustration, if you have an Osiris-enabled service that listens on two ports, the activator would open _two_, available, unprivileged ports.) The mapping you have proposed would be a simpler one than the one we use now. It would say "when we get a new connection on a given port, activate this corresponding deployment."

_Indeed, we already considered this very architecture at one point._ It's attractive in that it would obviate any need to sniff the connections to find information about the destination from the connection itself-- which is a somewhat complicated process. The reason we did _not_ use this approach is that getting the activator and the endpoints controller to _agree_ on which activator ports map to which ports of a given service is a major, major hurdle.

I will elaborate.

Upon scale-to-zero, with the architecture you have proposed, the endpoints controller, unlike today, would have to rewrite the service's endpoints to direct traffic to _specific_ ports on the activator. How would we get these two components to agree on _which_ ports?

To solve for this, the first thing to consider is: _Who_ decides which ports to use in the first place? The obvious answer is that the activator _must_ be the one to make this decision because _it_ knows which ports it currently has available to allocate for use with a given Osiris-enabled service. Two problems pop up immediately:

1. How does the activator inform the endpoints controller of which of its own ports it has allocated to each port of each Osiris-enabled service? I do not have a satisfactory answer for this yet.

1. What happens If the activator itself is scale to n > 1 replicas, how do you get all n replicas to independently allocate their ports _in the same way_? We wouldn't want, for instance, two activator instances where one has allocated port 5555 for service x port 80 activation while the _other_ has allocated port 7777 for the same purpose.

I will make no claim that these challenges are unsolvable. But they're not solvable without a lot of new complexity. _I also believe this is a very deep rabbit hole._ I haven't thought past the challenges above enough to predict what new problems will be introduced by the added complexity. There are, for sure, many, such as (but not limited to) : Are we going to let the activator modify its own `Service` and/or `Endpoints` resources? We'd pretty much have to make those endpoints discoverable and usable by Osiris-enabled services that are scaled to zero. But then what would happen, for instance, on `helm upgrade` when all of the activator's modifications to its own `Service` or `Endpoints` resources would be obliterated. 🤷‍♂ 

I'll reiterate that I don't think any of this is impossible. It's simply a rabbit hole that I do not have the bandwidth to fall down. If someone else wants to pursue this, I would certainly entertain a _detailed_ proposal and a PoC.

**Comment by ajbouh** [comment]
Thanks for the detailed write-up!

I don't understand the intricate interplay between the endpoints controller and the activator, but here's a an outline of something that might work without being too complex and without requiring too much coordination.

I say might because it depends on how much we can assume about clock drift.

Assigning a unique, bounded integer to every key in a set appears to be a solved problem in the world of open addressing (for hash tables). As long as each actor in the system uses the same hash function for the mapping, everything can work without additional coordination.

Of course, arriving at the same hash function is only possible if all actors have the same view of keys that require mapping.

One way to do this is to only handle services that have existed for some minimum quantity of time that is much greater than the propagation time for new kubernetes resources (for example resources updated on or before the second-to-last 5 minute mark). Then we can assume that all services will be mapped to the same offset in the available port range.

To avoid race conditions it would probably be best to support two different port ranges and alternate which is actually updated. This way it is always unambiguous which "version" of the map the activator is expected to use.

There are of course still open questions around how to modify the service to point to the activator, but I'm unfortunately ignorant to how the problem is solved in the current implementation.

**Comment by krancour** [comment]
> I don't understand the intricate interplay between the endpoints controller and the activator

There is no interplay. Both components work independently of one another.

As to the rest of what you propose...

I've concluded through my own due diligence that we cannot support plain TCP connections without incurring an undesired level of complexity and I'll ask you to trust that I didn't arrive at that lightly. Apart from new complexity being undesired, I've previously noted I do not have the bandwidth to implement such a solution or to mitigate the new failure modes that will arise from that complexity.

I recognize that your suggestions are aimed at reducing the complexity to the point where the solution becomes tenable and I cannot emphasize enough that your ideas _are welcome_. At the same time, please do recognize that "it _might_ work if if we do x, but I'm not sure", while well-intentioned, shifts the burden back to me to either explore your suggestion further or explain in detail why it cannot work. With my bandwidth being what it is, that's not a sustainable model for pursuing this further.

If you are serious about pursuing this, my best suggestion is that you take the time to understand the existing architecture, take the time to explore the unknowns in what you are suggesting, paying special attention to new failure modes and race conditions that may be created, then come back with a detailed proposal that leaves no stone unturned. Ideally, working code is even better.

**Comment by ajbouh** [comment]
Thanks for clarifying your perspective here.

Apologies for wasting your time. In my experience it's considered poor form to show up with a design *and* implementation without first soliciting input from the existing maintainer(s).

At the very least my comments here were intended to help plant seeds for a future implementation that might support this use case. I think the design I've outlined is both simple and correct as-is. I couldn't find any documented guarantees about worst-case-time-to-scale-down or about how existing edge cases are handled, so I made my caveat regarding clock-slew explicit.

Are there any docs that outline the existing architecture and  assumptions for the current implementation?  E.g. how is the existing system expected to handle event notifications falling behind? My searching of the current repository didn't yield much along these lines.

I'm motivated to help with implementation and testing, but I can't see it succeeding without support along the way from the existing maintainer(s).

Based on the conversation above, perhaps the project is currently too early/bandwidth constrained to support external contributors making major architectural changes? I can certainly relate to that sort of limitation.

Please correct me if any of the above is incorrect. Either way, I appreciate your quick and thorough responses.

**Comment by krancour** [comment]
> Apologies for wasting your time.

No apologies necessary. No time was wasted. And again, I cannot emphasize enough the ideas are appreciated.

> In my experience it's considered poor form to show up with a design and implementation without first soliciting input from the existing maintainer(s).

Certainly true. Coming to the table to with an implementation without some discussion occurring first could result in you wasting a lot of time on something that won't be accepted. That is in no one's interest. Let's say a design alone (i.e. without implementation) is a good starting point, but note that "if we do x it _might_ work" falls short of what might reasonably be considered a design.

> I think the design I've outlined is both simple and correct as-is.

I think the root issue here is that you haven't actually convinced me of this.

> I couldn't find any documented guarantees about worst-case-time-to-scale-down

Not entirely sure how this is relevant. In any event "scale down" is achieved by setting desired replicas to zero. Kubernetes' own replication controller takes it from there. Your only guarantees are whatever _it_ guarantees to you. But again... not sure why this is relevant.

> so I made my caveat regarding clock-slew explicit

tbqh, I didn't (and still don't) understand the caveat about clock skew. I especially don't understand it as it relates to "worst-case-time-to-scale-down." And I don't understand how _either_ of those things relate to port allocation on the activator. It's possible I am overlooking some obvious things here. I guess I have to own up to possibly having prematurely dismissed this idea due to a perception that you haven't connected the dots for me. Having said that, clearly connected dots are a reasonable expectation when major architectural changes are proposed.

> Are there any docs that outline the existing architecture and assumptions for the current implementation?

The readme explains the architecture at a high level. For greater understanding, admittedly, you've got to get into the code. The code is, at least, liberally commented. I'm sorry that I don't have a more satisfying answer for this.

> how is the existing system expected to handle event notifications falling behind?

Like the rest of Kubernetes, eventual consistency is the name of the game. Consider, for instance, what would happen to Kubernetes' own endpoints controller if it were not promptly notified that a certain pod was no longer in a ready state. The effect would be that traffic might be routed to an endpoint that isn't actually available. The effect on _our_ endpoints controller would be identical. Or consider, for instance, what would happen to an ingress controller if it were not promptly notified of changes to a service that it uses as a back end for an `Ingress`. The effect would be misdirected traffic. The effect on our activator would be the same. I'm afraid I really do not have a good answer for how a component can compensate for the delayed delivery of an event it doesn't even know it's expecting. This isn't an Osiris problem. It's a Kubernetes problem.

> Based on the conversation above, perhaps the project is currently too early/bandwidth constrained to support external contributors making major architectural changes? I can certainly relate to that sort of limitation.

Fair to say both. This _is_ early-- note there hasn't even been an alpha release. We've made this public, but never formally released anything with even minimal guarantees of any kind. Anyone using this is using edge code at their own risk. At the same time, interest from the community has been, shall we say, minimal. With maintainers having other responsibilities as well, it's fair to say this project isn't highly prioritized right now... this does, however, mean there is an opportunity for community members to step up and make big contributions...

It was never my intention to take the wind out of your sails. I hope I haven't and I apologize if I have. I can only re-iterate that ideas are welcome, but for a suggestion _of this scope_ (if I were sizing this, it is an epic), your proposal needs to be informed by the current architecture, framed in that context, and have all the dots explicitly connected. If that is something you can come back with, I welcome it.
